pkg/etw/eventdata.go

```diff
@@ -0,0 +1,23 @@
+package etw
+
+import (
+	"bytes"
+	"encoding/binary"
+)
+
+// EventData maintains a buffer which builds up the data for an ETW event. It
+// needs to be paired with EventMetadata which describes the event.
+type EventData struct {
+	buffer bytes.Buffer
+}
+
+// NewEventData returns a new EventData with an empty buffer.
+func NewEventData() *EventData {
+	return &EventData{}
+}
+
+// AddString appends the data for a string to the end of the buffer.

```

Comment:
There are some more complicated types that require separately specifying data and metadata.

For instance, to write an array of ints, you only do one call to write an array metadata, but have several calls to write each piece of data. That's a simple example, but once you start having types like nested structs and arrays of structs, it's useful to keep data and metadata apart.

I'm not opposed to adding some sort of helper for common types, in addition to the lower-level interface. I have been planning on waiting a little to see how the use cases for this code develop before putting that in.

Reply:
I think the important feedback here is that since this is a public package for others to consume, you should focus on making the exposed interface ergonomic and supportable. You can always expose the lower-level interface later, but exposing it now might make it harder to change the underlying implementation later.

E.g. for your int array example, I would guess most (all?) callers would rather just pass an int slice and let the underlying code handle the loop. This is less flexible since it requires that the caller have all their data available in a contiguous slice, but it's what TraceLogging.h already offers and that seems to be enough for most of Windows.

Reply:
I've done some work to figure out more of what the API shape should be. It still needs to be fleshed out a bit more, but the code in sample.go demonstrates the difference between the low-level and high-level API. I anticipate most uses being able to consume the high-level API, and have ported the Logrus hook to use it as well.